jaxtyping/_decorator.py

```diff
@@ -309,6 +310,29 @@ def __init__(self, *args, **kwargs):
             # in which case make a best-effort attempt to add shape information for any
             # type errors.
 
+            # we want to detect generators, and ignore return annotations on them,
+            # to avoid issues with O(n) typechecking trying to typecheck yielded values
+            wrp = fn
+            while hasattr(wrp, "__wrapped__"):
+                wrp = wrp.__wrapped__
+
+            if inspect.isgeneratorfunction(wrp) or inspect.isasyncgenfunction(wrp):
+                # recursively parse all the annotations, and mark all the jaxtyping
+                # annotations as not needing instance checks, while still being
+                # visible as original ones for the typechecker
+                def modify_annotation(ann):
+                    if isinstance(ann, _MetaAbstractArray):
+                        setattr(ann, "_skip_instancecheck", True)
+                    elif hasattr(ann, "__args__"):
+                        for sub_ann in get_args(ann):
+                            modify_annotation(sub_ann)
```

Comment:
I think the `hasattr` check can be removed: `get_args` will always return a tuple. (If necessary an empty tuple.)